### PR TITLE
presto-druid changed

### PR DESCRIPTION
**Pull request by mqiang**
presto-druid  changed

### CODE REVIEW

**Comment by mqiang** [comment]
1.Escape special fields such as 'count' when exec 'select * from table'
2.solve groupByFields can not find.
3.handle float filed UnsupportedOperationException
4.add presto-druid into tar.gz when packing

File: presto-server/pom.xml [hunk 1]
```diff
@@ -283,6 +283,15 @@
             <type>zip</type>
             <scope>provided</scope>
         </dependency>
+        
+        
+        <dependency>
+            <groupId>com.facebook.presto</groupId>
```
Comment (zhenxiao) [inline presto-server/pom.xml:289]: change the commit message to "add presto-druid connector to server tarball"

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidBrokerPageSource.java [hunk 2]
```diff
@@ -117,7 +117,7 @@ public Page getNextPage()
             JsonNode rootNode;
             try {
                 rootNode = OBJECT_MAPPER.readTree(data);
-                checkArgument(rootNode.isArray(), "broker Druid query should return Json Array");
+                checkArgument(rootNode.isArray(), "broker Druid query should return Json Array. detail error: "+data);
```
Comment (zhenxiao) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidBrokerPageSource.java:120]: nice catch.
I would suggest "broker Druid query should return Json Array:  " + data

Comment (zhenxiao) [reply presto-druid/src/main/java/com/facebook/presto/druid/DruidBrokerPageSource.java:120]: I would suggest only keep error message fix and float type fix in this commit, and change the commit message to: "Fix float value for druid connector"

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGeneratorContext.java [hunk 3]
```diff
@@ -262,7 +273,7 @@ private boolean hasAggregation()
 
         if (!groupByColumns.isEmpty()) {
             String groupByExpression = groupByColumns.stream()
-                    .map(expression -> selections.containsKey(expression) ? selections.get(expression).getDefinition() : expression.getName())
+                    .map(expression -> allSelections.containsKey(expression) ? allSelections.get(expression).getDefinition() :expression.getName())
```
Comment (zhenxiao) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGeneratorContext.java:276]: format: need a " " before expression

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java [hunk 4]
```diff
@@ -236,7 +236,12 @@ public DruidQueryGeneratorContext visitProject(ProjectNode node, DruidQueryGener
                         variable,
                         new Selection(druidExpression.getDefinition(), druidExpression.getOrigin()));
             });
-            return context.withProject(newSelections);
+
+            Map<VariableReferenceExpression, Selection> allSelections = context.getSelections().entrySet().stream()
```
Comment (zhenxiao) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java:240]: am not convinced by this change. for queries like:
select count(1) from table group by cityName

user need to specify cityName explicitly:
select cityName, count(1) from table group by cityName

We might not need to add cityName for user.

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java [hunk 5]
```diff
@@ -262,7 +262,7 @@ public DruidQueryGeneratorContext visitTableScan(TableScanNode node, DruidQueryG
             node.getOutputVariables().forEach(outputColumn -> {
                 DruidColumnHandle druidColumn = (DruidColumnHandle) (node.getAssignments().get(outputColumn));
                 checkArgument(druidColumn.getType().equals(DruidColumnHandle.DruidColumnType.REGULAR), "Unexpected druid column handle that is not regular: " + druidColumn);
-                selections.put(outputColumn, new Selection(druidColumn.getColumnName(), TABLE_COLUMN));
+                selections.put(outputColumn, new Selection("\\\"" + druidColumn.getColumnName() + "\\\"", TABLE_COLUMN));
```
Comment (zhenxiao) [inline presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java:265]: add a comment "escape column name with reserve words, e.g. column 'count'"

File: presto-druid/src/test/java/com/facebook/presto/druid/TestDruidQueryGenerator.java [hunk 6]
```diff
@@ -54,7 +54,9 @@ private void testDQL(
             String expressions = planNode.getOutputVariables().stream().map(v -> outputVariables.get(v.getName())).filter(v -> v != null).collect(Collectors.joining(", "));
             expectedDQL = expectedDQL.replace("__expressions__", expressions);
         }
-        assertEquals(druidQueryGeneratorResult.getGeneratedDql().getDql(), expectedDQL);
+        String generateDQL = druidQueryGeneratorResult.getGeneratedDql().getDql();
+        generateDQL = generateDQL.replaceAll("\\\\\"","");
```
Comment (zhenxiao) [inline presto-druid/src/test/java/com/facebook/presto/druid/TestDruidQueryGenerator.java:58]: put the same comment here

File: presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java (hunk 5 above)
Comment (zhenxiao) [reply presto-druid/src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java:265]: change commit message to be "escape keyword for druid column name"

**Comment by zhenxiao** [comment]
BTW, @mqiang could you please make sure all commits are under your same github account? Or, both of the accounts need to sign EasyCLA

**Comment by zhenxiao** [comment]
please fix code style:
[INFO] There are 5 errors reported by Checkstyle 8.16 with src/checkstyle/presto-checks.xml ruleset.
[ERROR] src/main/java/com/facebook/presto/druid/DruidQueryGeneratorContext.java:[276,126] (whitespace) WhitespaceAround: ':' is not followed by whitespace.
[ERROR] src/main/java/com/facebook/presto/druid/DruidBrokerPageSource.java:[120,112] (whitespace) WhitespaceAround: '+' is not followed by whitespace.
[ERROR] src/main/java/com/facebook/presto/druid/DruidBrokerPageSource.java:[120,112] (whitespace) WhitespaceAround: '+' is not preceded with whitespace.
[ERROR] src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java:[241] (regexp) RegexpMultiline: Multiple consecutive blank lines
[ERROR] src/test/java/com/facebook/presto/druid/TestDruidQueryGenerator.java:[58,54] (whitespace) WhitespaceAfter: ',' is not followed by whitespace.

**Comment by mqiang** [comment]
> please fix code style:
> [INFO] There are 5 errors reported by Checkstyle 8.16 with src/checkstyle/presto-checks.xml ruleset.
> [ERROR] src/main/java/com/facebook/presto/druid/DruidQueryGeneratorContext.java:[276,126] (whitespace) WhitespaceAround: ':' is not followed by whitespace.
> [ERROR] src/main/java/com/facebook/presto/druid/DruidBrokerPageSource.java:[120,112] (whitespace) WhitespaceAround: '+' is not followed by whitespace.
> [ERROR] src/main/java/com/facebook/presto/druid/DruidBrokerPageSource.java:[120,112] (whitespace) WhitespaceAround: '+' is not preceded with whitespace.
> [ERROR] src/main/java/com/facebook/presto/druid/DruidQueryGenerator.java:[241] (regexp) RegexpMultiline: Multiple consecutive blank lines
> [ERROR] src/test/java/com/facebook/presto/druid/TestDruidQueryGenerator.java:[58,54] (whitespace) WhitespaceAfter: ',' is not followed by whitespace.

@zhenxiao Thanks for your advice, i submit a new pull request. https://github.com/prestodb/presto/pull/14438

**Pull request closed without merging**